File: matlab/codes/noisemodel.m
function model = noisemodel(noisetype, param, dim)
    %function to generate different noise models. For now only global symmetric noise is supported.
    %TODO add  i) Local depolarising noise
    %Inputs:
    % noisetype : string specifying the noise model type.
    % param     : parameter specifying the strength of the noise.
    % dim       : dimension of the Dicke subspace which will be N+1. Where N is the number of qubits.
    %Outputs: 
    % model     : cell array of Kraus operators representing the noise model.
    switch noisetype
        case "globalsymmetric"
            model = globalsymmetric(param, dim); 
        case "tracepreserving"
            model = tracepreserving(param, dim);
        case "collectdeph"
            model = collective_dephasing(param, dim);  
        case "collectdeph_new"
            model = collective_dephasing_new(param, dim);    
        case "localampdamp"
            model = local_ampdamp(param, dim);      
        otherwise
            error('Unsupported noise model type. The noise models supported at the moment are: "globalsymmetric", "tracepreserving" and "ampdamp".');   
    end  
    
    function model = globalsymmetric(param, dim)
        persistent cache_dim cache_Ex cache_Ey cache_Ez cache_s
        if isempty(cache_dim) || cache_dim ~= dim 
            n = dim - 1;
            s = n / 2;
            Esm = zeros(dim);
            % Generate the Jx, Jy and Jz operators in the Dicke subspace. Note that this is a
            %  generalisation of the depolarising channel to higher dimensions.
            for j = -s:s-1
                Esm(j+s+1, j+s+2) = sqrt(s*(s+1) - j*(j+1));
            end
            Esp = Esm';
            Ex = (Esm + Esp) / 2;
            Ey = (Esm - Esp) / (2*1i);
            Ez = (Esp*Esm - Esm*Esp) / 2;
            cache_dim = dim;
            cache_Ex = Ex;
            cache_Ey = Ey;
            cache_Ez = Ez;
            cache_s = s; 
        end   
        p = param; % noise parameter
        E0 = ((1 - 0.5*(p*cache_s*(cache_s+1)))) * eye(dim);
        E1 = sqrt(p) * cache_Ex;
        E2 = sqrt(p) * cache_Ey;
        E3 = sqrt(p) * cache_Ez;
        model = {E0, E1, E2, E3};            
    end 

    function model = tracepreserving(param, dim)
            % Trace-preserving global symmetric noise model
            persistent cache_dim cache_Ex cache_Ey cache_Ez cache_s
            if isempty(cache_dim) || cache_dim ~= dim 
                n = dim - 1;
                s = n / 2;
                Esm = zeros(dim);
                for j = -s:s-1
                    Esm(j+s+1, j+s+2) = sqrt(s*(s+1) - j*(j+1));
                end
                Esp = Esm';
                Ex = (Esm + Esp) / 2;
                Ey = (Esm - Esp) / (2*1i);
                Ez = (Esp*Esm - Esm*Esp) / 2;
                cache_dim = dim;
                cache_Ex = Ex;
                cache_Ey = Ey;
                cache_Ez = Ez;
                cache_s = s;    
            end 
            p = param; % noise parameter
            E0 = sqrt(1 - p * cache_s *(cache_s + 1)) * eye(cache_dim);
            E1 = sqrt(p) * cache_Ex;
            E2 = sqrt(p) * cache_Ey;
            E3 = sqrt(p) * cache_Ez;
            model = {E0, E1, E2, E3};
        end 

    function K = collective_dephasing(param, dim)

        if dim < 1 || floor(dim) ~= dim
            error('dim must be a positive integer.');
        end

        N = dim - 1;
        J = N / 2;

        if J > 0
            maxParam = 1 / (2 * J * (J + 1));
            if param < 0 || param > maxParam
                error('param must be in [0, 1/(2*J*(J+1))] = [0, %.3g] for N = %d.', maxParam, N);
            end
        else
            if param ~= 0
                error('For dim = 1 (N = 0), param must be 0.');
            end
        end

        Sz = zeros(dim);
        for k = 0:N
            m = k - J;
            Sz(k+1, k+1) = m;
        end

        Sm = zeros(dim);
        for k = 1:N
            amp = sqrt(k * (N - k + 1));
            Sm(k, k+1) = amp;
        end

        Sp = Sm';

        H = Sz^2 + Sp*Sm + Sm*Sp;
        I = eye(dim);

        K0 = sqrtm(I - param * H);
        K1 = sqrt(param) * Sz;
        K2 = sqrt(param) * Sm;
        K3 = sqrt(param) * Sp;

        K = {K0, K1, K2, K3};

    end 

    function K = collective_dephasing_new(param, dim)

        N = dim - 1;
        J = N / 2;

        if param < 0 || param > 1/(2 * J * (J + 1))
            error('param must be in [0, 1/(2*J*(J+1))] for N = %d.', N);
        end

        Jz = zeros(dim);
        for k = 0:N
            m = k - J;
            Jz(k+1, k+1) = m;
        end

        Jp = zeros(dim);
        for k = 0:N-1
            amp = sqrt((k+1) * (N - k));
            Jp(k+2, k+1) = amp;
        end
        Jm = Jp';

        K1 = sqrt(param)      * Jp;
        K2 = sqrt(param)      * Jm;
        K3 = sqrt(2 * param)  * Jz;
        K4 = sqrt(1 - 2*param*J*(J+1)) * eye(dim);

        K = {K1, K2, K3, K4};

    end 

    function Kraus = local_ampdamp(param, dim)
        gamma = param; 
        % Generate the Kraus operators for local amplitude damping noise on N qubits.
        if ~(isscalar(dim) && dim == round(dim) && dim >= 2)
            error('dim must be an integer >= 2 (and equal to 2^N).');
        end
        if ~(isscalar(gamma) && isreal(gamma) && gamma >= 0 && gamma <= 1)
            error('gamma must be a real scalar in [0,1].');
        end
        
        N = dim; %number of qubits  
        %N = round(log2(dim));
        %if 2^N ~= dim
        %    error('dim must be exactly a power of 2 (dim = 2^N).');
        %end

        % ---- single-qubit amplitude damping Kraus ops ----
        E0 = [1 0; 0 sqrt(1-gamma)];
        E1 = [0 sqrt(gamma); 0 0];
        E  = {E0, E1};  % E{1}=E0, E{2}=E1

        % ---- build N-qubit Kraus ops ----
        numK = 2^N;
        Kraus = cell(1, numK);

        for s = 0:(numK-1)
            bits = dec2bin(s, N) - '0'; % length-N, leftmost bit is MSB
            K = 1;
            for q = 1:N
                K = kron(K, E{bits(q)+1});
            end
            Kraus{s+1} = K;
        end
    end
end

File: python/codes/noisemodel.py
import numpy as np
import qutip
from qutip.piqs.piqs import*
from qutip import*
from typing import List, Union
import qutip.piqs as piqs


def noisemodel(
    noise_type: str,
    num_qubits: int,
    gamma: float,
    dt: float,
    return_rep: str = "kraus",  # "kraus" | "choi" | "super"
    dynamics: str = "exact",    # "exact" | "approx"
):
    """
    Generate a quantum noise channel in PIQS and return either Kraus operators,
    the Choi matrix, or the superoperator.

    Parameters
    ----------
    noise_type : str
        Supported: 'global symmetric depolarizing', 'local symmetric depolarizing', 'global symmetric amplitude damping', 'local symmetric amplitude damping'
    num_qubits : int
        Number of qubits.
    gamma : float
        Noise rate parameter.
    dt : float
        Time step.
    return_rep : str
        'kraus' -> list of Kraus operators (slow: Choi eigen-decomposition)
        'choi'  -> Choi matrix (fast)
        'super' -> superoperator
    dynamics : str
        'exact'  -> use the exact PIQS Lindbladian dynamics exp(L t)
        'approx' -> use the second-order Kraus approximation from Eq. 16.
                    Currently only supported for global symmetric amplitude damping.

    Returns
    -------
    If return_rep == 'kraus': list[qutip.Qobj]
    If return_rep == 'choi' : qutip.Qobj
    If return_rep == 'super': qutip.Qobj
    """
    noise_type = noise_type.lower().strip()
    return_rep = return_rep.lower().strip()
    dynamics = dynamics.lower().strip()

    if noise_type == "global symmetric depolarizing":
        _require_exact_dynamics(noise_type, dynamics)
        return _global_symmetric_depolarizing(num_qubits, gamma, dt, return_rep)

    if noise_type == "local symmetric depolarizing":
        _require_exact_dynamics(noise_type, dynamics)
        return _local_symmetric_depolarizing(num_qubits, gamma, dt, return_rep)

    if noise_type == "global symmetric amplitude damping": 
        return _global_symmetric_amplitude_damping(num_qubits, gamma, dt, return_rep, dynamics)

    if noise_type == "local symmetric amplitude damping":
        _require_exact_dynamics(noise_type, dynamics)
        return _local_symmetric_amplitude_damping(num_qubits, gamma, dt, return_rep)

    raise ValueError(
        "Unsupported noise type. Supported types: "
        "'global symmetric depolarizing', 'local symmetric depolarizing', "
        "'global symmetric amplitude damping', 'local symmetric amplitude damping'."
    )


def _require_exact_dynamics(noise_type: str, dynamics: str):
    if dynamics != "exact":
        raise ValueError(
            f"dynamics='{dynamics}' is only implemented for "
            "'global symmetric amplitude damping'. "
            f"Use dynamics='exact' for '{noise_type}'."
        )


def _return_from_kraus(kraus: List[qutip.Qobj], return_rep: str):
    if return_rep == "kraus":
        return kraus
    if return_rep == "choi":
        return qutip.kraus_to_choi(kraus)
    if return_rep == "super":
        return qutip.kraus_to_super(kraus)
    raise ValueError("return_rep must be 'kraus', 'choi', or 'super'.")


def _global_symmetric_amplitude_damping_approx_kraus(
    num_qubits: int,
    gamma: float,
    dt: float,
) -> List[qutip.Qobj]:
    """
    Eq. 16 second-order Kraus approximation for global symmetric amplitude damping.

    The small parameter is alpha = gamma * dt and J_- is QuTiP/PIQS's
    collective lowering operator in the Dicke reduced basis. K3 is implemented
    literally as the Eq. 16 residual operator, so the map is trace-preserving up
    to the retained second-order dynamics rather than exactly CPTP at finite dt.
    """
    alpha = gamma * dt
    if alpha < 0:
        raise ValueError("gamma * dt must be non-negative for amplitude damping.")

    Jm = piqs.jspin(num_qubits, "-", basis="dicke")
    Jp = Jm.dag()
    ident = qutip.qeye(Jm.dims[0])
    Nop = Jp * Jm

    K0 = ident - (alpha / 2) * Nop + (alpha**2 / 8) * (Nop * Nop)
    K1 = np.sqrt(alpha) * Jm - (alpha**1.5 / 4) * (Jm * Nop + Nop * Jm)
    K2 = (alpha / np.sqrt(2)) * (Jm * Jm)

    K3 = ident - K0.dag() * K0 - K1.dag() * K1 - K2.dag() * K2

    return [K0, K1, K2, K3]


def _global_symmetric_depolarizing(num_qubits: int, gamma: float, dt: float, return_rep: str):
    """Global symmetric depolarizing-like noise using PIQS collective rates."""
    system = Dicke(num_qubits)
    system.collective_emission  = gamma
    system.collective_pumping   = gamma
    system.collective_dephasing = gamma

    L = system.liouvillian()
    E = (L * dt).expm()  # superoperator for the CPTP map

    if return_rep == "super":
        return E

    choi = to_choi(E)

    if return_rep == "choi":
        return choi

    if return_rep == "kraus":
        d = int(np.sqrt(L.shape[0]))  # since L is d^2 x d^2
        return _kraus_from_choi(choi, d_out=d, d_in=d)

    raise ValueError("return_rep must be 'kraus', 'choi', or 'super'.")


def _local_symmetric_depolarizing(num_qubits: int, gamma: float, dt: float, return_rep: str):
    """Local symmetric depolarizing-like noise using PIQS local rates."""
    system = Dicke(num_qubits)
    system.emission  = gamma
    system.pumping   = gamma
    system.dephasing = gamma

    L = system.liouvillian()
    E = (L * dt).expm()

    if return_rep == "super":
        return E

    choi = to_choi(E)

    if return_rep == "choi":
        return choi

    if return_rep == "kraus":
        d = int(np.sqrt(L.shape[0]))
        return _kraus_from_choi(choi, d_out=d, d_in=d)

    raise ValueError("return_rep must be 'kraus', 'choi', or 'super'.")

def _global_symmetric_amplitude_damping(
    num_qubits: int,
    gamma: float,
    dt: float,
    return_rep: str,
    dynamics: str = "exact",
): 
    """ generates a superoperator corresponding to Lindbladian with jump operator J_ = J_x - iJ_y, where Jx and Jy are the collective spin operators of N+1 dimensions. """
    if dynamics == "approx":
        kraus = _global_symmetric_amplitude_damping_approx_kraus(num_qubits, gamma, dt)
        return _return_from_kraus(kraus, return_rep)

    if dynamics != "exact":
        raise ValueError("dynamics must be either 'exact' or 'approx'.")

    system = Dicke(num_qubits)
    system.collective_emission = gamma
    L = system.liouvillian()
    E = (L * dt).expm()    
    if return_rep == "super":
        return E 
    if return_rep == "choi":
        choi = to_choi(E) 
        return choi
    if return_rep == "kraus": 
        d = int(np.sqrt(L.shape[0]))
        choi = to_choi(E)
        return _kraus_from_choi(choi, d_out = d, d_in = d)
    raise ValueError("return_rep must be 'kraus', 'choi', or 'super'.")

def _local_symmetric_amplitude_damping(num_qubits: int, gamma: float, dt: float, return_rep: str):
    """ generates a superoperator corresponding to Lindbladian with jump operators J_-^(i) = sigma_-^(i) for i=1,...,N, where sigma_-^(i) is the lowering operator on the i-th qubit. """
    system = Dicke(num_qubits)
    system.emission = gamma
    L = system.liouvillian()
    E = (L * dt).expm()    
    if return_rep == "super":
        return E 
    choi = to_choi(E)
    if return_rep == "choi": 
        return choi
    if return_rep == "kraus": 
        d = int(np.sqrt(L.shape[0]))
        return _kraus_from_choi(choi, d_out = d, d_in = d)
    raise ValueError("return_rep must be 'kraus', 'choi', or 'super'.")


def _kraus_from_choi(choi: qutip.Qobj, d_out: int, d_in: int, tol: float = 1e-15) -> List[qutip.Qobj]:
    """
    Convert a Choi matrix to a list of Kraus operators using eigen-decomposition.
    This is the slow step for large dimensions.

    Kraus operators returned have shape (d_out x d_in).
    """
    choi = (choi + choi.dag()) / 2  # enforce Hermiticity
    w, v = np.linalg.eigh(choi.full())

    Ks: List[qutip.Qobj] = []
    for k in range(len(w)):
        if w[k] <= tol:
            continue
        vec = np.sqrt(w[k]) * v[:, k]
        Kmat = vec.reshape((d_out, d_in), order="F")  # column-stacking
        Ks.append(qutip.Qobj(Kmat, dims=[[d_out], [d_in]]))

    return Ks
